### ISSUE TITLE
Preferred Serialization and Canonical encoding in CBOR

### ISSUE THREAD

**Issue by Sekenre** [agronholm/cbor2#28]
The CBOR mailing list has been discussing the definition of Canonical in the standard and have been making changes. I wanted to document these and maybe discuss how they might be implemented in cbor2.

Updated draft standard: https://datatracker.ietf.org/doc/draft-ietf-cbor-7049bis

What constitutes a true canonical encoding can be redefined at will by protocol implementers, CBOR standard provides guidelines.

A general encoder/decoder like cbor2 will need to support a number of variations and validate them.

Constraints which have been discussed on the ietf mailing list and in the updated draft:

* Shortest possible float representations
* Fixed float representations
* Fixed integer representations
* Maps sorted by lexicographical ordering of encoded value (DRAFT)
* Maps sorted by ascending length then lexicographic (RFC7049)

There may be tagged arrays created for fixed length binary encodings of float values. (Tag values TBD)

See: https://datatracker.ietf.org/doc/draft-ietf-cbor-array-tags/

Decoders may need to validate these by raising errors if the following conditions are met:

* Indefinite length types
* Floating point values not in shortest form
* Floating point values not in fixed representation
* Integers not in shortest form
* Integers not in Fixed form
* Unsorted maps
* Maps sorted with the wrong algorithm
* Maps with duplicate keys
* Incorrect tag type

Instead of a single `canonical=True` argument there needs to be separate flags for each potential constraint.

For example, if a device expects only 16bit floating point data you could create the encoder like this:

```python
encoder = CBOREncoder(f, float_format="binary16")
encoder.encode(data)
```

Or for a minimal float encoding and sorted maps using the encoded length

```python
encoder = CBOREncoder(f, float_format="minimal", sort_maps=True, sort_by_length=True)
encoder.encode(data)
```

On the decoding side:

```python
decoder = CBORDecoder(f, validate_floats_as="binary16")
result = decoder.decode()
```

```python
decoder = CBORDecoder(f, validate_floats_as="minimal",
    validate_map_order=True,
    ordered_by_length=True,
    ignore_duplicate_keys=False)
result = decoder.decode()
```

Of course these argument names and the way they are set up are just intended as an example.

**Comment by agronholm** [comment]
At this point I think you should become the maintainer of cbor2. How about it?

**Comment by Sekenre** [comment]
I'd be honoured! I am inexperienced but very keen :grinning:

**Comment by fsssosei** [comment]
The code seems to be a long integer time complexity O (n ^ 2)?
Let's say I have a 10,000 bit integer

**Comment by agronholm** [comment]
What code where?

**Comment by Sekenre** [comment]
> The code seems to be a long integer time complexity O (n ^ 2)?
> Let's say I have a 10,000 bit integer

Hi @fsssosei could you open a new issue for this and I will look at it when I get the chance?

**Comment by henryk** [comment]
To add some food for thought: Here is an article discussing the different "canonical" encodings in CBOR: https://www.imperialviolet.org/2022/04/17/canonsofcbor.html
It also proposes a naming scheme. RFC 7049 seems to describe "three-step" ordering (but could be read ambigously), RFC 8949 describes "one-step" ordering.

Best as I can tell, cbor2 currently implements three-step ordering. For starters, the documentation could point out the different ways a "canonical" CBOR can be canonical, and document the current state in the library.

**Comment by Sekenre** [comment]
Absolutely. Thanks @henryk for bringing that article to my attention. I have been playing around with splitting the canonical settings into their own options, i.e. the 3 options for map ordering, fixed or variable-sized floating point. Etc. Then have a backwards compatible default.

Of course it's easy in python, harder in C. I don't think I will implement validating whether something is canonical, but will document how someone could do so.

**Comment by rohanmahy** [comment]
Well, cbor2 is using the more complicated RFC 7049 canonical sort order.

If I have a dict like this:

```
{
  cbor2.CBORSimpleValue(59): 1,
  501: 1
}
```

RFC 7049 says the canonical order sorts by length first and therefore outputs:
```
A2  # map(2)
  F8 3B    # primitive(59)
  01          # unsigned(1)
  19 01F5 # unsigned(501)
  01          # unsigned(1)
```

while RFC 8949 says to order the keys lexicographical, so the integer key comes first:
A2  # map(2)
  19 01F5 # unsigned(501)
  01          # unsigned(1)
  F8 3B    # primitive(59)
  01          # unsigned(1)
```

it looks like `encode_canonical_map` is the place where this happens in `cbor2/_encoder.py`

**Comment by rohanmahy** [comment]
A simpler example would be:

```
{
   257: True,
   "a": True
}
```

The integer key should sort before the text key in RFC 8494 regardless of length, whereas the shorter key is sorted first in RFC 7094
